### ISSUE TITLE
Incorrect outputs.

### ISSUE THREAD

**Issue by vsantjr** [spyder-ide/spyder#14073]
### What steps will reproduce the problem?

An MNIST char and posterior function plot are wrongly merged together, if a Run the whole program. If I run cell by cell, they are ok.

## Versions

* Spyder version: 4.1.5 
* Python version: 3.7.6
* Qt version: 5.12.9
* PyQt5 version: 5.12.3
* Operating System: Darwin 18.2.0

### Dependencies

```

# Mandatory:
applaunchservices >=0.1.7      :  0.2.1 (OK)
atomicwrites >=1.2.0           :  1.4.0 (OK)
chardet >=2.0.0                :  3.0.4 (OK)
cloudpickle >=0.5.0            :  1.6.0 (OK)
diff_match_patch >=20181111    :  20200713 (OK)
intervaltree                   :  None (OK)
IPython >=4.0                  :  7.12.0 (OK)
jedi =0.17.1                   :  0.17.1 (OK)
nbconvert >=4.0                :  5.6.1 (OK)
numpydoc >=0.6.0               :  1.1.0 (OK)
parso =0.7.0                   :  0.7.0 (OK)
pexpect >=4.4.0                :  4.8.0 (OK)
pickleshare >=0.4              :  0.7.5 (OK)
psutil >=5.3                   :  5.7.2 (OK)
pygments >=2.0                 :  2.5.2 (OK)
pylint >=1.0                   :  2.6.0 (OK)
pyls >=0.34.0;<1.0.0           :  0.35.1 (OK)
qdarkstyle >=2.8               :  2.8.1 (OK)
qtawesome >=0.5.7              :  1.0.1 (OK)
qtconsole >=4.6.0              :  4.6.0 (OK)
qtpy >=1.5.0                   :  1.9.0 (OK)
sphinx >=0.6.6                 :  3.2.1 (OK)
spyder_kernels >=1.9.4;<1.10.0 :  1.9.4 (OK)
watchdog                       :  None (OK)
zmq >=17                       :  18.1.1 (OK)

# Optional:
cython >=0.21                  :  None (OK)
matplotlib >=2.0.0             :  3.1.3 (OK)
numpy >=1.7                    :  1.18.1 (OK)
pandas >=0.13.1                :  1.0.1 (OK)
scipy >=0.17.0                 :  1.4.1 (OK)
sympy >=0.7.3                  :  1.5.1 (OK)
```

**Comment by vsantjr** [comment]
These are the parts of the code. 

----

from matplotlib import pyplot as plt
import numpy as np

plt.close('all')
print('x_train shape before: ', x_train[90].shape)
plt.imshow(x_train[90].reshape((28, 28)), cmap="gray")
print('x_train shape after: ', x_train[90].shape)

....
....

def log_softmax(x):
    return x - x.exp().sum(-1).log().unsqueeze(-1)

def model(xb):
    return log_softmax(xb @ weights + bias)



#%%
a = np.linspace(0, 100, 101)
print(a)
at = torch.from_numpy(a)
print(at)

plt.plot(at, log_softmax(at), label='log_soft')
plt.show()

**Comment by ccordoba12** [comment]
Hey @vsantjr, thanks for reporting. Please post a complete code example so we can run it in our side. Unfortunately, the code you posted above is incomplete.

**Comment by vsantjr** [comment]
Hi @ccordoba12,

Here it is code. It is one from PyTorch tutorial, with some minor changes. I use python 3.7. Running all the code, the two images appear as one, and incorrectly merged (in my case).

Thanks.

---- 

#!/usr/bin/env python3
# -*- coding: utf-8 -*-

#%%

from pathlib import Path
import requests

DATA_PATH = Path("data")
PATH = DATA_PATH / "mnist"

PATH.mkdir(parents=True, exist_ok=True)

URL = "http://deeplearning.net/data/mnist/"
FILENAME = "mnist.pkl.gz"

if not (PATH / FILENAME).exists():
        content = requests.get(URL + FILENAME).content
        (PATH / FILENAME).open("wb").write(content)
        

#%%
import pickle
import gzip

with gzip.open((PATH / FILENAME).as_posix(), "rb") as f:
        ((x_train, y_train), (x_valid, y_valid), _) = pickle.load(f, encoding="latin-1")
        


#%%
from matplotlib import pyplot as plt
import numpy as np

plt.close('all')
print('x_train shape before: ', x_train[90].shape)
plt.imshow(x_train[90].reshape((28, 28)), cmap="gray")
print('x_train shape after: ', x_train[90].shape)
        


#%%
import torch

x_train, y_train, x_valid, y_valid = map(
    torch.tensor, (x_train, y_train, x_valid, y_valid)
)
n, c = x_train.shape
x_train, x_train.shape, y_train.min(), y_train.max()
print(x_train, y_train)




#%%
import math

weights = torch.randn(784, 10) / math.sqrt(784)
 
weights.requires_grad_()


bias = torch.zeros(10, requires_grad=True)


#%%
def log_softmax(x):
    return x - x.exp().sum(-1).log().unsqueeze(-1)

def model(xb):
    return log_softmax(xb @ weights + bias)





#%%
a = np.linspace(0, 100, 101)
print(a)
at = torch.from_numpy(a)
print(at)



plt.plot(at, log_softmax(at), label='log_soft')
plt.show()

**Comment by ccordoba12** [comment]
Thanks for your help @vsantjr!

@impact27, could you take a look at this one?

**Comment by impact27** [comment]
You should open a new figure with plt.figure() for each figure you use. Otherwise the merged figure is the expected behaviour

**Comment by vsantjr** [comment]
Thanks @impact27. 
See the code below. After some tests, I conclude that I only need to open a new figure with plt.figure() if I am displaying data as an image (imshow). I can use as many as plt.plot, plt.show() in sequence without requiring any plt.figure(). If this is the expected behaviour, ok then.

#!/usr/bin/env python3
# -*- coding: utf-8 -*-

#%%

from pathlib import Path
import requests

DATA_PATH = Path("data")
PATH = DATA_PATH / "mnist"

PATH.mkdir(parents=True, exist_ok=True)

URL = "http://deeplearning.net/data/mnist/"
FILENAME = "mnist.pkl.gz"

if not (PATH / FILENAME).exists():
        content = requests.get(URL + FILENAME).content
        (PATH / FILENAME).open("wb").write(content)
        

#%%
import pickle
import gzip

with gzip.open((PATH / FILENAME).as_posix(), "rb") as f:
        ((x_train, y_train), (x_valid, y_valid), _) = pickle.load(f, encoding="latin-1")
        


#%%
from matplotlib import pyplot as plt
import numpy as np

plt.close('all')
print('x_train shape before: ', x_train[90].shape)
plt.imshow(x_train[90].reshape((28, 28)), cmap="gray")
print('x_train shape after: ', x_train[90].shape)
        


#%%
import torch

x_train, y_train, x_valid, y_valid = map(
    torch.tensor, (x_train, y_train, x_valid, y_valid)
)
n, c = x_train.shape
x_train, x_train.shape, y_train.min(), y_train.max()
print(x_train, y_train)




#%%
import math

weights = torch.randn(784, 10) / math.sqrt(784)
 
weights.requires_grad_()


bias = torch.zeros(10, requires_grad=True)


#%%
def log_softmax(x):
    return x - x.exp().sum(-1).log().unsqueeze(-1)

def model(xb):
    return log_softmax(xb @ weights + bias)





#%%
a = np.linspace(0, 100, 101)
print(a)
at = torch.from_numpy(a)
print(at)


plt.figure() # I must do it
plt.plot(at, log_softmax(at), label='log_soft')
plt.show()

#%%
# no plt.figure here 
plt.plot(at, np.sin(at), label='sine')
plt.show()

print('After some modifications!')

"""
 Just some simple code
"""
#%%

print('New image')

plt.imshow(x_train[120].reshape((28, 28)), cmap="gray")


#%%
b = np.linspace(0, 50, 51)
print(b)

plt.figure() # I must do it   
plt.plot(b, b**2, label='squared')
plt.show()    

#%%
# no plt.figure here   
plt.plot(b, np.cos(b), label='cos')
plt.show()

**Comment by impact27** [comment]
So to be precise, what happens is:
if you plot something with `plt.plot`, `plt.imshow` or any plotting function without creating a figure, pyplot creates one for you. All subsequent plotting function will go on the same figure until you create a new one.

If you use `plt.show`, you show all the figure and "close" them if you use the default plotting backend (not if you have an interactive one), so the next plotting function will create a new figure.
